pkg/trace/api/telemetry_test.go

```diff
@@ -0,0 +1,204 @@
+	assert := assert.New(t)
+
+	srv := assertingServer(t, func(req *http.Request, body []byte) error {
+		assert.Equal("body", string(body), "invalid request body")
+		assert.Equal("test_apikey", req.Header.Get("DD-API-KEY"))
+		assert.Equal("test_hostname", req.Header.Get("DD-Agent-Hostname"))
+		assert.Equal("test_env", req.Header.Get("DD-Agent-Env"))
+		assert.Equal("/path", req.URL.Path)
+
+		atomic.AddUint64(&endpointCalled, 1)
+		return nil
+	})
+	defer mockConfigMap(map[string]interface{}{
+		"api_key":                     "test_apikey",
+		"apm_config.telemetry.dd_url": srv.URL,
+		"hostname":                    "test_hostname",
+		"skip_ssl_validation":         true,
+		"env":                         "test_env",
+	})() // reset config after the test
+
+	req, rec := newRequestRecorder(t)
+	recv := NewHTTPReceiver(loadConfig(t), nil, nil, nil)
+	recv.buildMux().ServeHTTP(rec, req)
+
+	assert.Equal("OK", recordedResponse(t, rec))
+	assert.Equal(uint64(1), atomic.LoadUint64(&endpointCalled))
+}
+
+func TestTelemetryProxyMultipleEndpoints(t *testing.T) {
+	var endpointCalled uint64
+	assert := assert.New(t)
+
+	mainBackend := assertingServer(t, func(req *http.Request, body []byte) error {
+		assert.Equal("body", string(body), "invalid request body")
+		assert.Equal("test_apikey_1", req.Header.Get("DD-API-KEY"))
+		assert.Equal("test_hostname", req.Header.Get("DD-Agent-Hostname"))
+		assert.Equal("test_env", req.Header.Get("DD-Agent-Env"))
+
+		atomic.AddUint64(&endpointCalled, 2)
+		return nil
+	})
+	additionalBackend := assertingServer(t, func(req *http.Request, body []byte) error {
+		assert.Equal("body", string(body), "invalid request body")
+		assert.Equal("test_apikey_2", req.Header.Get("DD-API-KEY"))
+		assert.Equal("test_hostname", req.Header.Get("DD-Agent-Hostname"))
+		assert.Equal("test_env", req.Header.Get("DD-Agent-Env"))
+
+		atomic.AddUint64(&endpointCalled, 3)
+		return nil
+	})
+
+	additionalEndpoints := make(map[string]string)
+	additionalEndpoints[additionalBackend.URL+"/"] = "test_apikey_2"
+	// proxy must ignore malformed urls
+	additionalEndpoints["111://malformed_url.example.com"] = "test_apikey_3"
+
+	defer mockConfigMap(map[string]interface{}{
+		"apm_config.telemetry.additional_endpoints": additionalEndpoints,
+		"apm_config.telemetry.dd_url":               mainBackend.URL,
+		"api_key":                                   "test_apikey_1",
+		"hostname":                                  "test_hostname",
+		"skip_ssl_validation":                       true,
+		"env":                                       "test_env",
+	})()
+
+	req, rec := newRequestRecorder(t)
+	recv := NewHTTPReceiver(loadConfig(t), nil, nil, nil)
+	recv.buildMux().ServeHTTP(rec, req)
+
+	assert.Equal("OK", recordedResponse(t, rec))
+
+	// because we use number 2,3 both endpoints must be called to produce 5
+	// just counting number of requests could give false results if first endpoint
+	// was called twice
+	if atomic.LoadUint64(&endpointCalled) != 5 {
```

Comment:
![Tux, the Linux mascot](https://c.tenor.com/dp_hQBGT0rIAAAAC/think-smart.gif)

Reply:
Maths FTW :)